src/main/java/uk/gov/ons/ctp/response/casesvc/message/CaseReceiptReceiver.java

```diff
@@ -1,15 +1,81 @@
+
+import com.godaddy.logging.Logger;
+import com.godaddy.logging.LoggerFactory;
+import java.sql.Timestamp;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.integration.annotation.MessageEndpoint;
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
 import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.response.casesvc.domain.model.Case;
+import uk.gov.ons.ctp.response.casesvc.domain.model.CaseEvent;
 import uk.gov.ons.ctp.response.casesvc.message.feedback.CaseReceipt;
+import uk.gov.ons.ctp.response.casesvc.message.feedback.InboundChannel;
+import uk.gov.ons.ctp.response.casesvc.representation.CategoryDTO;
+import uk.gov.ons.ctp.response.casesvc.service.CaseService;
 
 /** The reader of CaseReceipts from queue */
-public interface CaseReceiptReceiver {
+@MessageEndpoint
+public class CaseReceiptReceiver {
+  private static final Logger log = LoggerFactory.getLogger(CaseReceiptReceiver.class);
+
+  private static final String EXISTING_CASE_NOT_FOUND = "No existing case found";
+
+  @Autowired private CaseService caseService;
+
   /**
    * To process CaseReceipts read from queue
    *
    * @param caseReceipt to process
-   * @throws CTPException when case state transition error
+   * @throws CTPException CTPException
    */
-  void process(CaseReceipt caseReceipt) throws CTPException;
+  @Transactional(propagation = Propagation.REQUIRED, readOnly = false, value = "transactionManager")
+  @ServiceActivator(inputChannel = "caseReceiptTransformed", adviceChain = "caseReceiptRetryAdvice")
+  public void process(CaseReceipt caseReceipt) throws CTPException {
+    log.with("case_receipt", caseReceipt).debug("entering process with caseReceipt");
+    UUID caseId = UUID.fromString(caseReceipt.getCaseId());
+    InboundChannel inboundChannel = caseReceipt.getInboundChannel();
+    Timestamp responseTimestamp =
+        new Timestamp(caseReceipt.getResponseDateTime().toGregorianCalendar().getTimeInMillis());
+
+    Case existingCase = caseService.findCaseById(caseId);
+    log.with("existing_case", existingCase).debug("Found existing case");
+
+    CategoryDTO.CategoryName category = null;
+    switch (inboundChannel) {
+      case OFFLINE:
+        category = OFFLINE_RESPONSE_PROCESSED;
+        break;
+      case ONLINE:
+        category = ONLINE_QUESTIONNAIRE_RESPONSE;
+        break;
+      case PAPER:
+        category = PAPER_QUESTIONNAIRE_RESPONSE;
+        break;
+      default:
+        break;
+    }
+
+    if (existingCase == null) {
+      log.with("case_id", caseId).error(EXISTING_CASE_NOT_FOUND);
+    } else {
+      CaseEvent caseEvent = new CaseEvent();
+      caseEvent.setCaseFK(existingCase.getCasePK());
+      caseEvent.setCategory(category);
```

Comment:
if category is not on ethe 3 listed will be null then ?

Reply:
yes. I don't like it much either. I would like to go through a lot of the Client and Service classes and get rid of the "return null instead of throwing an error" pattern, which I don't think is very good or the right way to do things. Code copied from CaseReceiptReceiverImpl.java and is unchanged.